### PR TITLE
[flang] Pass to add frame pointer attribute

### PR DESCRIPTION
**Pull request by Radu2k**
Pass to add frame pointer attribute in Flang

### CODE REVIEW

**Comment by kiranchandramohan** [comment]
You have some clang-format issues.

**Review by jeanPerier — COMMENTED** [review]
A few comments inline. My more generic questions are:
- Can the equivalent be achieved by passing the option to LLVM codegen instead of decorating all functions?
- The pass as it is today is not very flang specific, should it belong to the Func or MLIR LLVM dialect. Unless there is a point for MLIR optimization to know about these aspect, why does it have to be set up in the chain?

**Comment by Radu2k** [comment]
@Dinistro @gysit Do you think that adding this pass as an LLVM Func pass would be acceptable?

File: flang/include/flang/Tools/CLOptions.inc [hunk 1]
```diff
@@ -311,6 +311,10 @@ inline void createDefaultFIRCodeGenPassPipeline(
   if (config.VScaleMin != 0)
     pm.addPass(fir::createVScaleAttrPass({config.VScaleMin, config.VScaleMax}));
 
+  fir::FunctionAttrTypes functionAttrs;
+  functionAttrs.framePointerKind = config.FramePointerKind;
+  pm.addPass(fir::createFunctionAttrPass(functionAttrs));
```
[the remark this answers is not in the document]
Comment (Radu2k) [reply flang/include/flang/Tools/CLOptions.inc:316]: As in Clang, FramePointerKind is usually set, even if it is not specified, therefore the current logic accommodates this dynamic. In flang/include/flang/Tools/CrossToolHelpers.h it is either updated or set to None. As well the FunctionAttrs is meant to be reused for multiple function attributes that require to be added in LLVM IR.

**Comment by jeanPerier** [comment]
> > * Can the equivalent be achieved by passing the option to LLVM codegen instead of decorating all functions?
Would that be adding a conversion pattern for func.func in the Flang Codegen.cpp?

No, I meant later: i.e. in the actual LLVM IR processing that flang calls in FrontendActions.cpp.

llc has a `--frame-pointer` option, so there is likely a way to pass this option to the llvm codegen as a global option instead of setting it on all function at the MLIR level if the only goal is to propagate it to llvm backend.

File: flang/include/flang/Optimizer/Transforms/Passes.td [hunk 2]
```diff
@@ -349,4 +349,23 @@ def VScaleAttr : Pass<"vscale-attr", "mlir::func::FuncOp"> {
   let constructor = "::fir::createVScaleAttrPass()";
 }
 
+def FunctionAttr : Pass<"function-attr", "mlir::func::FuncOp"> {
+  let summary = "Enhance functions with different attributes";
```
Comment (tblah) [inline flang/include/flang/Optimizer/Transforms/Passes.td:353]:
```suggestion
  let summary = "Set attributes on all functions";
```

File: flang/lib/Frontend/CompilerInvocation.cpp [hunk 3]
```diff
@@ -245,6 +245,18 @@ static void parseCodeGenArgs(Fortran::frontend::CodeGenOptions &opts,
 
   opts.AliasAnalysis = opts.OptimizationLevel > 0;
 
+  if (const llvm::opt::Arg *a =
+          args.getLastArg(clang::driver::options::OPT_mframe_pointer_EQ)) {
+    llvm::StringRef s = a->getValue();
+    assert(s == "none" || s == "non-leaf" || s == "all");
```
Comment (tblah) [inline flang/lib/Frontend/CompilerInvocation.cpp:251]: I don't think this should be an assertion
1) Yes users shouldn't use `flang -fc1`, but it is still kind of a user interface. Crashing with a stack trace looks like a bug rather than a typo in the command line options.
2) Assertions are not enabled on all builds

It would be better to print an error or warning message

File: flang/lib/Optimizer/Transforms/FunctionAttr.cpp [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+#include "mlir/IR/Matchers.h"
+#include "mlir/IR/TypeUtilities.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "mlir/Transforms/RegionUtils.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/raw_ostream.h"
+
+#include <algorithm>
```
Comment (tblah) [inline flang/lib/Optimizer/Transforms/FunctionAttr.cpp:38]: What are you using `<algorithm>` for?

File: flang/lib/Optimizer/Transforms/FunctionAttr.cpp [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+namespace fir {
+#define GEN_PASS_DECL_FUNCTIONATTR
```
Comment (tblah) [inline flang/lib/Optimizer/Transforms/FunctionAttr.cpp:41]: I don't think you need this line (see how most other flang passes only request the definition, not the declaration in the .cpp file).

File: flang/lib/Optimizer/Transforms/FunctionAttr.cpp [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+} // namespace
+
+void FunctionAttrPass::runOnOperation() {
+  LLVM_DEBUG(llvm::dbgs() << "=== Begin " DEBUG_TYPE " ===\n");
+  mlir::func::FuncOp func = getOperation();
+
+  LLVM_DEBUG(llvm::dbgs() << "Func-name:" << func.getSymName() << "\n");
+
+  auto context = &getContext();
```
Comment (tblah) [inline flang/lib/Optimizer/Transforms/FunctionAttr.cpp:67]: Please spell out `auto`.

In general, we only use `auto` in contexts where the type is obvious from the rest of the line. For example
```
auto attr = mlir::LLVM::FramePointerKindAttr::get(...); // type is obviously an mlir::LLVM::FramePointerKindAttr
auto func = dyn_cast<mlir::func::FuncOp>(op); // type is obviously an mlir::func::FuncOp
```

Here it is not obvious what the type is, and that makes the `&` especially hard to reason about.

**Comment by jeanPerier** [comment]
> llc has a --frame-pointer option, so there is likely a way to pass this option to the llvm codegen as a global option instead of setting it on all function at the MLIR level if the only goal is to propagate it to llvm backend.

I just discussed a bit with @vzakhari and one advantage of setting it per function is to later be able to merge llvm IR files compiled with different options. That would explain why clang does it. I would still add nothing if the option is None since this is the default, but seems OK to me to decorate the functions as you are doing based on this argument.

File: flang/lib/Optimizer/Transforms/FunctionAttr.cpp (hunk 5 above)
Comment (tblah) [reply flang/lib/Optimizer/Transforms/FunctionAttr.cpp:41]: Yeah I was thinking of moving it to Passes.h. But you're right, if we don't need to make the declaration public then we shouldn't.

File: flang/lib/Frontend/CompilerInvocation.cpp [hunk 7]
```diff
@@ -248,13 +248,21 @@ static void parseCodeGenArgs(Fortran::frontend::CodeGenOptions &opts,
...
+    if (s == "none") {
       opts.setFramePointer(llvm::FramePointerKind::None);
-    else if (s == "non-leaf")
-      opts.setFramePointer(llvm::FramePointerKind::NonLeaf);
-    else
-      opts.setFramePointer(llvm::FramePointerKind::All);
+    } else {
+      if (s == "non-leaf")
+        opts.setFramePointer(llvm::FramePointerKind::NonLeaf);
+      else
```
Comment (tblah) [inline flang/lib/Frontend/CompilerInvocation.cpp:263]: Sorry I didn't spot this before. It will match strings other than `"all"`. We should provide an error on an invalid input instead of silently setting it to all.

For example, imagine if Clang added a new frame pointer kind called "apple". Somebody might try to set the frame pointer to "apple" in flang, and they wouldn't get any error telling them it was not supported, but flang would actually set the frame pointer kind to "all": probably not doing what the user wanted.

File: flang/lib/Optimizer/Transforms/FunctionAttr.cpp [hunk 8]
```diff
@@ -64,32 +40,19 @@ void FunctionAttrPass::runOnOperation() {
 
   LLVM_DEBUG(llvm::dbgs() << "Func-name:" << func.getSymName() << "\n");
 
-  auto context = &getContext();
-
-  func->setAttr("frame_pointer", mlir::LLVM::FramePointerKindAttr::get(
-                                     context, framePointerKind));
+  mlir::MLIRContext* context = &getContext();
```
Comment (tblah) [inline flang/lib/Optimizer/Transforms/FunctionAttr.cpp]: nit
```
mlir::MLIRContext *context = &getContext();
```

File: flang/test/Driver/fixed-free-detection.f90 [hunk 9]
```diff
@@ -25,5 +25,5 @@
 ! MULTIPLEFORMS:program freeform
 ! MULTIPLEFORMS-NOT:programfixedform
 ! MULTIPLEFORMS-NEXT:end
-! MULTIPLEFORMS-NEXT:programfixedform
+! MULTIPLEFORMS:programfixedform
```
Comment (tblah) [inline flang/test/Driver/fixed-free-detection.f90]: Why this change?

File: flang/test/Driver/func-attr.f90 [hunk 10]
```diff
@@ -6,7 +6,7 @@
 subroutine func
 end subroutine func
 
-! CHECK-NONEFP: attributes #0 = { "frame-pointer"="none" }
+! CHECK-NONEFP: attributes #0 = {{{.*}}}
```
Comment (tblah) [inline flang/test/Driver/func-attr.f90:9]: Instead of `{{.*}}` could you check that there are no attributes here? This way we check that no frame pointer attributes are added for `None`. `{{.*}}` can match anything.

**Review by tblah — COMMENTED** [review]
Thanks for the updates

File: flang/lib/Frontend/CompilerInvocation.cpp [hunk 11]
```diff
@@ -248,13 +248,21 @@ static void parseCodeGenArgs(Fortran::frontend::CodeGenOptions &opts,
...
+        opts.setFramePointer(llvm::FramePointerKind::All);
```
[the remark this answers is not in the document]
Comment (Radu2k) [reply flang/lib/Frontend/CompilerInvocation.cpp:264]: This is a change suggested above by Kiran. 
https://github.com/llvm/llvm-project/pull/74598#discussion_r1417479590

[the remark this answers is not in the document]
Comment (tblah) [reply flang/lib/Frontend/CompilerInvocation.cpp:264]: Ahh I missread the old side of the diff. What I liked would be something more like
```
if (s == "option1")
  ...
else if (s == "option2")
  ...
else if (s == "option3")
  ...
else
  error!
```

There is no need for braces on any of these because there will only be a single line in the body of the if statements.

Kiran was right that in this case
```
if (cond)
  something
else
  if (cond2)
     foo
  else
    bar
```

It is a good idea to put braces around the outer if because otherwise it is harder to see to which `if` the final `else` belongs.

[the remark this answers is not in the document]
Comment (kiranchandramohan) [reply flang/lib/Frontend/CompilerInvocation.cpp:264]: I agree with @tblah that they should all be at the same indentation. The old formattting (copied below) made it look like there are two levels of indentation and hence my comment. 
```
    if (s == "none")
      opts.setFramePointer(llvm::FramePointerKind::None);
    else
      if (s == "non-leaf")
        opts.setFramePointer(llvm::FramePointerKind::NonLeaf);
      else
        opts.setFramePointer(llvm::FramePointerKind::All);
 ```

File: flang/test/Driver/fixed-free-detection.f90 (hunk 9 above)
Comment (Radu2k) [reply flang/test/Driver/fixed-free-detection.f90]: This was done in order to restore the code back to default. This is due to the frame pointer value being specified as argument instead of modifying the test.

File: flang/lib/Frontend/CompilerInvocation.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (Radu2k) [reply flang/lib/Frontend/CompilerInvocation.cpp:264]: Ok, I will then update this in the following commit.

File: flang/include/flang/Tools/CLOptions.inc [hunk 12]
```diff
@@ -310,10 +311,17 @@ inline void createDefaultFIRCodeGenPassPipeline(
...
+  
+  // Add function attributes 
   fir::FunctionAttrTypes functionAttrs;
-  functionAttrs.framePointerKind = config.FramePointerKind;
-  pm.addPass(fir::createFunctionAttrPass(functionAttrs));
+  
+  if (config.FramePointerKind != llvm::FramePointerKind::None) {  
+    if (config.FramePointerKind == llvm::FramePointerKind::NonLeaf) 
+        functionAttrs.framePointerKind = mlir::LLVM::framePointerKind::FramePointerKind::NonLeaf;
+    else functionAttrs.framePointerKind = mlir::LLVM::framePointerKind::FramePointerKind::All;
```
[the remark this answers is not in the document]
Comment (Radu2k) [reply flang/include/flang/Tools/CLOptions.inc]: Why do consider this is extra?

[the remark this answers is not in the document]
Comment (tblah) [reply flang/include/flang/Tools/CLOptions.inc]: Ahh never mind, I missed the conversion from an llvm frame pointer to an mlir frame pointer.

Why use an llvm frame pointer at all? We could just store the mlir frame pointer from the start and not need to convert

File: flang/test/Driver/fixed-free-detection.f90 (hunk 9 above)
Comment (Radu2k) [reply flang/test/Driver/fixed-free-detection.f90]: This change was made when the frame-pointer had as the default value None, therefore setting it to None, impacted the test. How, I can not recall, but since that is no longer the implementation I will remove any changes in this file, thanks for pointing it out.

File: flang/include/flang/Tools/CLOptions.inc (hunk 12 above)
[the remark this answers is not in the document]
Comment (Radu2k) [reply flang/include/flang/Tools/CLOptions.inc]: To optimise the code I have removed any invocation from the FrontendActions.cpp. The parameter is declared in `CodeGedOptions.def`, set in the `CompilerInvocation.cpp` based on the user requirements and called `FramePointerKind = opts.getFramePointer();` when `MLIRToLLVMPassPipelineConfig` is initialised.

In both `CodeGedOptions.def` and `CompilerInvocation.cpp` I saw that llvm namespace is predominantly used, and did not used the mlir FramePointerKind.

**Review by tblah — APPROVED** [review]
Thanks for the update. LGTM

**Comment by kiranchandramohan** [comment]
> > llc has a --frame-pointer option, so there is likely a way to pass this option to the llvm codegen as a global option instead of setting it on all function at the MLIR level if the only goal is to propagate it to llvm backend.
> 
> I just discussed a bit with @vzakhari and one advantage of setting it per function is to later be able to merge llvm IR files compiled with different options. That would explain why clang does it. I would still add nothing if the option is None since this is the default, but seems OK to me to decorate the functions as you are doing based on this argument.

@jeanPerier just checking whether you are OK with this pass being here or would prefer it to be in LLVM dialect?

**Review by jeanPerier — COMMENTED** [review]
Thanks for the update, please take care of Kiran's comments. LGTM otherwise.

File: flang/lib/Frontend/CompilerInvocation.cpp [hunk 13]
```diff
@@ -245,6 +245,24 @@ static void parseCodeGenArgs(Fortran::frontend::CodeGenOptions &opts,
 
   opts.AliasAnalysis = opts.OptimizationLevel > 0;
 
+  if (const llvm::opt::Arg *a =
+          args.getLastArg(clang::driver::options::OPT_mframe_pointer_EQ)) {
+    llvm::StringRef s = a->getValue();
+
+    if (!(s == "none" || s == "non-leaf" || s == "all")) {
+      const auto debugWarningId = diags.getCustomDiagID(
+          clang::DiagnosticsEngine::Error, "Frame pointer: %0");
```
[the remark this answers is not in the document]
Comment (Radu2k) [reply flang/lib/Frontend/CompilerInvocation.cpp:254]: Just added, thanks for pointing it out!

File: flang/lib/Frontend/CompilerInvocation.cpp [hunk 14]
```diff
@@ -245,6 +245,24 @@ static void parseCodeGenArgs(Fortran::frontend::CodeGenOptions &opts,
...
+      diags.Report(debugWarningId).AddString(a->getValue());
+    }
+
+    if (s == "none")
+      opts.setFramePointer(llvm::FramePointerKind::None);
+    else if (s == "non-leaf")
+      opts.setFramePointer(llvm::FramePointerKind::NonLeaf);
+    else
+      opts.setFramePointer(llvm::FramePointerKind::All);
```
Comment (banach-space) [inline flang/lib/Frontend/CompilerInvocation.cpp:263]: 1. You don't need `else` in this block.
2. You could use `llvm::StringSwitch` and use the default value to mark a failure (as opposed to `if (!(s == "none" || s == "non-leaf" || s == "all"))` above)

File: flang/test/Driver/save-mlir-temps.f90 [hunk 15]
```diff
@@ -51,9 +51,9 @@
 ! Content to check from the MLIR outputs
 !--------------------------
 ! MLIR-FIR-NOT: llvm.func
-! MLIR-FIR: func.func @{{.*}}main() {
+! MLIR-FIR: func.func @{{.*}}main(){{.*}}{
```
Comment (banach-space) [inline flang/test/Driver/save-mlir-temps.f90:54]: What do these extra braces capture?

File: flang/lib/Frontend/CompilerInvocation.cpp (hunk 13 above)
Comment (banach-space) [inline flang/lib/Frontend/CompilerInvocation.cpp:254]: I couldn't find it - which one is it?

File: flang/test/Driver/save-mlir-temps.f90 (hunk 15 above)
Comment (Radu2k) [reply flang/test/Driver/save-mlir-temps.f90:54]: This was added due to failing the test when it was a default parameter. I do not see how  having `{{.*}}` instead of a whitespace should be an issue if no parameters are checked.

File: flang/include/flang/Optimizer/Transforms/Passes.td [hunk 16]
```diff
@@ -349,4 +349,24 @@ def VScaleAttr : Pass<"vscale-attr", "mlir::func::FuncOp"> {
   let constructor = "::fir::createVScaleAttrPass()";
 }
 
+def FunctionAttr : Pass<"function-attr", "mlir::func::FuncOp"> {
+  let summary = "This is a generic pass that adds function attributes expected at LLVM IR level";
```
Comment (kiranchandramohan) [inline flang/include/flang/Optimizer/Transforms/Passes.td:353]:
```suggestion
  let summary = "Pass that adds function attributes expected at LLVM IR level";
```

File: flang/include/flang/Optimizer/Transforms/Passes.td [hunk 17]
```diff
@@ -349,4 +349,24 @@ def VScaleAttr : Pass<"vscale-attr", "mlir::func::FuncOp"> {
...
+  let description = [{ This feature introduces a general attribute aimed at customizing function characteristics. 
+     Options include:
+     Add "frame-pointer" attribute to functions: Set an attribute for the frame 
+     pointer on functions, to avoid saving the frame pointer in a register in 
+     functions where it is unnecessary. 
+     This eliminates the need for instructions to save, establish, and restore 
+     frame pointers, while also freeing up an additional register in numerous functions. 
+     However, this approach can make debugging unfeasible on certain machines.
+  }];
```
Comment (kiranchandramohan) [inline flang/include/flang/Optimizer/Transforms/Passes.td:362]:
```suggestion
  let description = [{ This feature introduces a general attribute aimed at
     customizing function characteristics. 
     Options include:
     Add "frame-pointer" attribute to functions: Set an attribute for the frame 
     pointer on functions, to avoid saving the frame pointer in a register in 
     functions where it is unnecessary. This eliminates the need for
     instructions to save, establish, and restore frame pointers, while also
     freeing up an additional register in numerous functions. However, this
     approach can make debugging unfeasible on certain machines.
  }];
```

**Review by kiranchandramohan — APPROVED** [review]
LG. Please wait for @banach-space

File: flang/lib/Frontend/CompilerInvocation.cpp (hunk 13 above)
Comment (banach-space) [reply flang/lib/Frontend/CompilerInvocation.cpp]: An error will cause the compiler to crash - I can't see any `RUN` lines in [rame-pointer-forwarding.f90](https://github.com/llvm/llvm-project/pull/74598/files#diff-fad5151400801da3763ab30d3470d69a748b236f0793afc41a9a94f6432a7c78) that would cause the compiler to crash.

Also, please don't "resolve" threads that haven't been concluded yet - this makes reviewing harder.

File: flang/test/Driver/frame-pointer-forwarding.f90 [hunk 18]
```diff
@@ -1,9 +1,12 @@
 ! Test that flang-new forwards -fno-omit-frame-pointer and -fomit-frame-pointer Flang frontend
-! RUN: %flang -fno-omit-frame-pointer --target=x86-none-none -fsyntax-only -### %s -o %t 2>&1  | FileCheck %s
-! CHECK: "-mframe-pointer=all"
+! RUN: %flang --target=aarch64-none-none -fsyntax-only -### %s -o %t 2>&1  | FileCheck %s --check-prefix=CHECK-NOVALUE
+! CHECK-NOVALUE: "-mframe-pointer=non-leaf"
```
Comment (banach-space) [inline flang/test/Driver/frame-pointer-forwarding.f90:3]:
```suggestion
! CHECK-NOVALUE: "-fc1" "-mframe-pointer=non-leaf"
```

Same remark for other check lines.

File: flang/test/Driver/mlir-debug-pass-pipeline.f90
Comment (banach-space) [inline flang/test/Driver/mlir-debug-pass-pipeline.f90:1]: This change doesn't make sense to me. This file tests default pass pipelines for various debug levels. Why is `-fomit-frame-pointer` needed here?

File: flang/test/Driver/save-mlir-temps.f90 (hunk 15 above)
Comment (banach-space) [reply flang/test/Driver/save-mlir-temps.f90:54]: Could you explain why it was failing? I just don't see how this particular change is related to what this PR is implementing. It would help if you included code "before" and "after" this PR.

File: flang/lib/Frontend/CompilerInvocation.cpp (hunk 13 above)
Comment (kiranchandramohan) [reply flang/lib/Frontend/CompilerInvocation.cpp]: @banach-space the test is in func-attr.f90 https://github.com/llvm/llvm-project/pull/74598/files#diff-1d4cea32c615e677fc20e34ffd2f5c3e0f84b00f5af6a692d787747e17bd9e3d

```
! RUN: not %flang_fc1 -triple aarch64-none-none -mframe-pointer=wrongval -emit-llvm -o - %s 2>&1| FileCheck %s --check-prefix=CHECK-WRONGVALUEFP
```

File: flang/test/Driver/func-attr.f90 [hunk 19]
```diff
@@ -0,0 +1,23 @@
+! Test that -mframe-pointer can accept only specific values.
```
Comment (banach-space) [inline flang/test/Driver/func-attr.f90:1]: You seem to be testing a bit more in this file than this comment suggests (e.g. that `-mframe-pointer` will make the compiler insert the new attributes)

Please, could you:
* update the comment to capture what's actually being tested (the `-mframe-pointer` flag)
* rename the file - it's testing a particular flag rather than any "func"tion "attr"ibute

File: flang/test/Driver/mlir-debug-pass-pipeline.f90
Comment (Radu2k) [reply flang/test/Driver/mlir-debug-pass-pipeline.f90:1]: This is needed because depending on the target, if the frame pointer is not specified, then the best option will be automatically added, therefore a 

```
'func.func' Pipeline
  FunctionAttr
```
will appear or not, depending of the target, at the end of the file. 

Therefore it is mandatory to add omitting frame pointer.

File: flang/test/Driver/frame-pointer-forwarding.f90 (hunk 18 above)
Comment (Radu2k) [reply flang/test/Driver/frame-pointer-forwarding.f90:3]: If we add this then it will be added as `"-fc1"{{.*}}"`.  Why is it necessary?

File: flang/lib/Frontend/CompilerInvocation.cpp (hunk 13 above)
Comment (banach-space) [reply flang/lib/Frontend/CompilerInvocation.cpp]: Now I see :)

File: flang/test/Driver/save-mlir-temps.f90 (hunk 15 above)
Comment (Radu2k) [reply flang/test/Driver/save-mlir-temps.f90:54]: Since this is target dependent, frame pointer will be added or not. It is either adding `{{.*}}` or adding `-fomit-frame-pointer`.

File: flang/test/Driver/frame-pointer-forwarding.f90 (hunk 18 above)
Comment (banach-space) [reply flang/test/Driver/frame-pointer-forwarding.f90:3]: See the comment at the top of the file:

> ! Test that flang-new forwards -fno-omit-frame-pointer and -fomit-frame-pointer Flang frontend

In order to identify that this is indeed checking the **frontend** driver invocation (as per the comment, that's the intention), you need to check for `-fc1`. That's something that was missing here to begin with, so strictly speaking it's tangential to your change. But since you are updating these CHECK lines anyway ...

File: flang/test/Driver/mlir-debug-pass-pipeline.f90
Comment (banach-space) [reply flang/test/Driver/mlir-debug-pass-pipeline.f90:1]: Thanks for the explanation!

> Therefore it is mandatory to add omitting frame pointer

No. You can also update the expected output. So there are 2 options. And what would be preferred - this test is checking pass pipelines and including `-fomit-frame-pointer` suggests that that flag will impact those pipelines. That's not the case though.

You can do this:
```
! ALL-LABEL: Fortran::lower::VerifierPass
! ALL-NEXT: 'func.func' Pipeline
! ALL: InlineElementals
! ALL-NEXT: LowerHLFIROrderedAssignments
```

Comment (Radu2k) [reply flang/test/Driver/mlir-debug-pass-pipeline.f90:1]: That looks good to me, thought it needs to check line by line(flang/test/Driver/mlir-debug-pass-pipeline.f90 - lines 35-38) but will update, replacing ALL-NEXT to ALL at line 85.
